src/client.rs

```diff
@@ -363,14 +363,15 @@ impl NatsClient {
     /// Performs a request to the server following the Request/Reply pattern. Returns a future containing the MSG that will be replied at some point by a third party
     ///
     /// Returns `impl Future<Item = Message, Error = NatsError>`
-    pub fn request(
+    pub fn request_with_timeout(
```

Comment:
So, as far as the functionality of the timeout, everything is looking good. I do have a concern about the segmentation of this into two different methods.

- if we are going to use two different methods, then I think the `request_with_timeout` should probably not take the `timeout` as an `Option`, as the `_with_timeout` part of the name implies that a timeout is expected.
- if we do want to make it optional, then perhaps we can just augment the original `request` method to take an `Option`al timeout. This would be a backwards incompatible change to the lib, but it is currently at `0.1` ... so we could just do a `0.2` release, and that shouldn't be an issue. We could just update the changelog to describe the change.

Personally, I am biased towards one `request` method which takes an optional timeout. I think that interface is a bit more flexible. Thoughts?

Reply:
I agree, I did this not to create an unnecessary breaking change on the API for a feature such as this one.

What I can suggest: 
- Leaving `request` as-is. => `pub fn request(&self, subject: String, payload: Bytes)`
- Making the `timeout` parameter in `request_with_timeout` without the wrapping `Option` (so, mandatory. => `pub fn request_with_timeout(&self, subject: String, payload: Bytes, mut timeout: Duration)`
- Having an internal `fn` that contains the code, takes an optional timeout, and in which both functions would go. => `fn internal_request(&self, subject: String, payload: Bytes, mut timeout: Option<Duration>)`

What do you think?

Reply:
Totally your call. My main concern is just with the public interface though. I do think that having a single `request` method with an optional timeout is a bit more of a straightforward interface.

To be sure, I definitely understand the desire to not make a backwards incompatible change to the lib; however, making such a change while the library is at `0.1` is far better than needing to make such a change in the future, especially as we get closer to a `1.0` release. Either way, the call is yours and both options should work just fine. Go with what you feel `:)`